lib/asyncSender.ts

```diff
@@ -0,0 +1,181 @@
+
+import {
+   Delivery, Message, Sender as RheaSender, SessionEvents
+} from "rhea";
+import * as log from "./log";
+import { BaseSender, BaseSenderOptions } from "./sender";
+import { SenderEvents } from "rhea";
+import { OnAmqpEvent, EventContext } from "./eventContext";
+import { Session } from "./session";
+import { OperationTimeoutError } from "./operationTimeoutError";
+
+/**
+ * @internal
+ */
+export interface PromiseLike {
+  resolve: (value?: any) => void;
+  reject: (reason?: any) => void;
+  timer: NodeJS.Timer;
+}
+
+/**
+ * Describes the event listeners that can be added to the AsyncSender.
+ * @interface Sender
+ */
+export declare interface AsyncSender {
+  on(event: SenderEvents, listener: OnAmqpEvent): this;
+}
+
+export interface AsyncSenderOptions extends BaseSenderOptions {
+  /**
+   * The duration in which the promise to send the message should complete (resolve/reject).
+   * If it is not completed, then the Promise will be rejected after timeout occurs.
+   * Default: `20 seconds`.
+   */
+  messageTimeoutInSeconds?: number;
+}
+
+/**
+ * Describes the async version of the sender where one can await on the message being sent.
+ * @class AsyncSender
+ */
+export class AsyncSender extends BaseSender {
+  /**
+   * The duration in which the promise to send the message should complete (resolve/reject).
+   * If it is not completed, then the Promise will be rejected after timeout occurs.
+   * Default: `20 seconds`.
+   */
+  messageTimeoutInSeconds: number;
+  /**
+   * @property {Map<number, PromiseLike} deliveryDispositionMap Maintains a map of delivery of
+   * messages that are being sent. It acts as a store for correlating the dispositions received
+   * for sent messages.
+   */
+  deliveryDispositionMap: Map<number, PromiseLike> = new Map<number, PromiseLike>();
+
+  constructor(session: Session, sender: RheaSender, options: AsyncSenderOptions = {}) {
+    super(session, sender, options);
+    this.messageTimeoutInSeconds = options.messageTimeoutInSeconds || 20;
+
+    const onSuccess = (context: EventContext) => {
+      const id = context.delivery!.id;
+      if (this.deliveryDispositionMap.has(context.delivery!.id)) {
+        const promise = this.deliveryDispositionMap.get(id) as PromiseLike;
+        clearTimeout(promise.timer);
+        log.sender(
+          "[%s] Event: 'Accepted', Found the delivery with id %d in the map of " +
+          "sender '%s' on amqp session '%s' and cleared the timer.",
+          this.connection.id, id, this.name, this.session.id
+        );
+        const deleteResult = this.deliveryDispositionMap.delete(id);
+        log.sender(
+          "[%s] Event: 'Accepted', Successfully deleted the delivery with id %d from " +
+          "the map of sender '%s' on amqp session '%s' and cleared the timer: %s.",
+          this.connection.id, id, this.name, this.session.id, deleteResult
+        );
```

Comment:
Why not combine the 2 log statements above?
`clearTimeout` and `map.delete` are synchronous calls. It possible can't happen that one runs and not the other.

In the sea of logs, it will be one less type for each sent message

Reply:
True. Will combing them.